src/workflows/airqo_etl_utils/bigquery_api.py

```
@@ -766,6 +766,123 @@
+            pd.DataFrame: A DataFrame containing the most recent record for each unique identifier.
+
+        Raises:
+            google.api_core.exceptions.GoogleAPIError: If the query execution fails.
+        """
+
+        selected_columns = ", ".join(columns) if columns else "*"
+        query = f"""
+        SELECT {selected_columns}
+        FROM `{table}`
+        QUALIFY ROW_NUMBER() OVER (PARTITION BY {unique_id} ORDER BY {offset_column} DESC) = 1;
+        """
```

Ruff (0.12.2)

Possible SQL injection vector through string-based query construction

(S608)